### ISSUE TITLE
[IR] [Refactor] Support `return` in serial kernels

### ISSUE THREAD

**Issue by yuanming-hu** [taichi-dev/taichi#909]
## Remove ArgStoreStmt

https://github.com/taichi-dev/taichi/blob/f5373b1395a66c78506de0fd3c172dffe0d444d0/taichi/ir/ir.h#L985

Currently the only use of `ArgStoreStmt` is for returning values in a kernel. I personally feel like this is a confusing design (https://github.com/taichi-dev/taichi/pull/839#issuecomment-622244367). I suggest removing this statement and replace it with `KernelReturnStmt`.

## Add KernelReturnStmt
`KernelReturnStmt` can only work in `serial` offloaded tasks.
 - Set the return value;
 - Terminate the kernel. Actually, we should enforce that `KernelReturnStmt` can only be the final statement of a serial kernel. If we add this restriction, then no need to worry about terminating the kernel.

## Benefits
 1. SNode readers can finally use return
 2. We can allow return values of arbitrary kernels, e.g.
```python
@ti.kernel
def compute_sum():
  s = 0.0  
  for i in x:
    s += x[i]
  return s

s = compute_sum()
```

## Discussions
 - How about returning a vector/matrix?

**Comment by archibate** [comment]
Cool! This is so cool because we never use `ArgStoreStmt` in python-defined kernels. But removing `ArgStoreStmt`, what if we want multiple return values like the pythonic`return a, b`?

**Comment by yuanming-hu** [comment]
Good point. `KernelReturnStmt` should have a member `std::vector<Stmt *> return_values` to support multiple return values.

**Comment by archibate** [comment]
Does `return` make sense in diff-kernels? e.g.:
```
@ti.kernel
def func(x: ti.i32):
  return x ** 2

@ti.kernel
def func_grad(x: ti.i32):
  return 2 * x  # am I understanding correct?
```
